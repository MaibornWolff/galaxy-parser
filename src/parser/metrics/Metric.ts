/**
 * Interface for carrying the result of a metric calculation.
 */
export interface MetricResult {
    /**
     * Name of the metric.
     */
    metricName: string;

    /**
     * Value of the metric.
     */
    metricValue: number;
}

export interface Relationship {
    fromNamespace: string;
    toNamespace: string;
    fromSource: string;
    toSource: string;
    fromClassName: string;
    toClassName: string;
    usageType: string | "usage" | "extends" | "implements";
}

export interface CouplingMetrics {
    outgoing_dependencies: number;
    incoming_dependencies: number;
    coupling_between_objects: number;
    instability: number;
}

export interface CouplingResult {
    relationships: Relationship[];
    metrics: Map<string, CouplingMetrics>;
}

/**
 * Interface for calculating a concrete metric.
 */
export interface Metric {
    /**
     * Calculates the metric value for the supplied file.
     * @param parseFile Source code file for which the metric value should be calculated.
     * @return A MetricResult containing the calculated metric value.
     */
    calculate(parseFile: ParseFile): MetricResult;

    /**
     * Returns the name of this metric.
     * @return The name of this metric.
     */
    getName(): string;
}

export interface CouplingMetric {
    calculate(files: ParseFile[]): CouplingResult;
    getName(): string;
}

/**
 * Represents a file to be parsed, including its path and the file extension.
 */
export interface ParseFile {
    /**
     * File extension of the file.
     */
    fileExtension: string;

    /**
     * Path to the file.
     */
    filePath: string;
}
