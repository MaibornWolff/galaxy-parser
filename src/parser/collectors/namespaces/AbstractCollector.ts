import { QueryResolver } from "./resolver/QueryResolver";
import { FilenameResolver } from "./resolver/FilenameResolver";
import { ParseFile } from "../../metrics/Metric";

export interface NamespaceDefinition {
    namespace: string;
    className: string;
    classType: string | "interface" | "class";
    source: string;
    namespaceDelimiter: string;
    extendedClass?: string;
    implementedClasses: string[];
}

export enum NamespaceResolvingStrategy {
    Query,
    Filename,
}

export abstract class AbstractCollector {
    protected abstract getNamespaceResolvingStrategy(): NamespaceResolvingStrategy;
    protected abstract getNamespaceDelimiter(): string;
    protected abstract getNamespacesQuery(): string;

    getNamespaces(parseFile: ParseFile): Map<string, NamespaceDefinition> {
        if (this.getNamespaceResolvingStrategy() === NamespaceResolvingStrategy.Query) {
            return new QueryResolver().getNamespaces(
                parseFile,
                this.getNamespaceDelimiter(),
                this.getNamespacesQuery()
            );
        }

        if (this.getNamespaceResolvingStrategy() === NamespaceResolvingStrategy.Filename) {
            return new FilenameResolver().getNamespaces();
        }

        throw Error(
            "Unsupported Namespace Resolving Strategy " + this.getNamespaceResolvingStrategy()
        );
    }
}
