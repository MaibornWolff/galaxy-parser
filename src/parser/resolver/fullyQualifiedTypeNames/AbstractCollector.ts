import { type ParsedFile } from "../../metrics/Metric.js";
import { QueryStrategy } from "./resolverStrategy/QueryStrategy.js";
import { FilenameStrategy } from "./resolverStrategy/FilenameResolver.js";

export type FullyQTN = {
    namespace: string;
    className: string;
    classType: "interface" | "class";
    source: string;
    namespaceDelimiter: string;
    extendedClass?: string;
    implementedClasses: string[];
};

export enum NamespaceResolvingStrategy {
    Query,
    Filename,
}

export abstract class AbstractCollector {
    protected abstract getNamespaceResolvingStrategy(): NamespaceResolvingStrategy;
    protected abstract getNamespaceDelimiter(): string;
    protected abstract getNamespacesQuery(): string;

    getFullyQTNs(parsedFile: ParsedFile): Map<string, FullyQTN> {
        if (this.getNamespaceResolvingStrategy() === NamespaceResolvingStrategy.Query) {
            return new QueryStrategy().getFullyQTNs(
                parsedFile,
                this.getNamespaceDelimiter(),
                this.getNamespacesQuery(),
            );
        }

        if (this.getNamespaceResolvingStrategy() === NamespaceResolvingStrategy.Filename) {
            return new FilenameStrategy().getFullyQTNs();
        }

        throw new Error(
            "Unsupported Namespace Resolving Strategy " +
                this.getNamespaceResolvingStrategy().toString(),
        );
    }
}
