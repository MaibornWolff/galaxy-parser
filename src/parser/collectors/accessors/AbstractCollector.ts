import { NamespaceCollector } from "../NamespaceCollector";
import { ParseFile } from "../../metrics/Metric";
import { formatCaptures } from "../../helper/Helper";
import { QueryBuilder } from "../../queries/QueryBuilder";
import { grammars } from "../../helper/Grammars";
import { TreeParser } from "../../helper/TreeParser";
import { NamespaceDefinition } from "../namespaces/AbstractCollector";

export interface PublicAccessor {
    name: string;
    namespaces: NamespaceDefinition[];
    filePath: string;
    returnType: string;
}

export abstract class AbstractCollector {
    protected abstract getPublicAccessorsQuery(): string;

    getPublicAccessors(
        parseFile: ParseFile,
        namespacesOfFile: Map<string, NamespaceDefinition>
    ): Map<string, PublicAccessor[]> {
        const publicAccessors = new Map<string, PublicAccessor[]>();

        if (this.getPublicAccessorsQuery()) {
            const tree = TreeParser.getParseTree(parseFile);
            const queryBuilder = new QueryBuilder(grammars.get(parseFile.language), tree);
            queryBuilder.setStatements([this.getPublicAccessorsQuery()]);

            const publicAccessorsQuery = queryBuilder.build();
            const publicAccessorsCaptures = publicAccessorsQuery.captures(tree.rootNode);
            const accessorsTextCaptures: {
                name: string;
                text: string;
                usageType?: string;
                source?: string;
            }[] = formatCaptures(tree, publicAccessorsCaptures);

            console.log("public accessors captures", accessorsTextCaptures);

            // first index must be the return type
            // second index must be the accessor name
            // third index (not_empty_accessor flag) must be ignored, if present
            for (let index = 0; index < accessorsTextCaptures.length; index += 1) {
                const publicAccessor: PublicAccessor = {
                    name: "",
                    namespaces: Array.from(namespacesOfFile.values()),
                    filePath: parseFile.filePath,
                    returnType: accessorsTextCaptures[index].text,
                };

                index++;
                publicAccessor.name = accessorsTextCaptures[index].text;

                const accessorsByName = publicAccessors.get(publicAccessor.name);
                if (accessorsByName === undefined) {
                    publicAccessors.set(publicAccessor.name, []);
                }

                publicAccessors.get(publicAccessor.name)?.push(publicAccessor);

                if (accessorsTextCaptures[index + 1]?.name === "not_empty_accessor") {
                    index++;
                }
            }
        }

        return publicAccessors;
    }
}
