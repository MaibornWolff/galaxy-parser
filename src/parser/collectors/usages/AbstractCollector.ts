import { QueryBuilder } from "../../queries/QueryBuilder";
import { grammars } from "../../helper/Grammars";
import { formatCaptures } from "../../helper/Helper";
import { TreeParser } from "../../helper/TreeParser";
import { NamespaceReference } from "../namespaces/AbstractCollector";

export interface UsageReference {
    usedNamespace: string;
    sourceOfUsing: string;
    alias?: string;
    source: string;
}

export abstract class AbstractCollector {
    protected abstract getUsagesQuery(): string;

    /*
     * Results caching not needed at the moment
     */
    getUsages(parseFile: ParseFile, packages: Map<string, NamespaceReference>): UsageReference[] {
        const tree = TreeParser.getParseTree(parseFile);

        const queryBuilder = new QueryBuilder(grammars.get(parseFile.language), tree);
        queryBuilder.setStatements([this.getUsagesQuery()]);

        const usagesQuery = queryBuilder.build();
        const usagesCaptures = usagesQuery.captures(tree.rootNode);
        const usagesTextCaptures = formatCaptures(tree, usagesCaptures);

        console.log(usagesTextCaptures);

        const usagesOfFile: UsageReference[] = [];

        for (let index = 0; index < usagesTextCaptures.length; index++) {
            if (usagesTextCaptures[index].name === "namespace_use_alias") {
                const namespaceAlias = usagesTextCaptures[index].text;
                usagesOfFile[index - 1].alias = namespaceAlias;

                continue;
            }

            const namespaceName = usagesTextCaptures[index].text;
            usagesOfFile.push({
                usedNamespace: namespaceName,
                sourceOfUsing: parseFile.filePath,
                alias: "",
                source: parseFile.filePath,
            });
        }

        return usagesOfFile;
    }
}
