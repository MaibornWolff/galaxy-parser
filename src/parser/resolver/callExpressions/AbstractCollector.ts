import { ParseFile } from "../../metrics/Metric";
import { formatCaptures } from "../../helper/Helper";
import { QueryBuilder } from "../../queries/QueryBuilder";
import { TreeParser } from "../../helper/TreeParser";
import { FullyQTN } from "../fullyQualifiedTypeNames/AbstractCollector";
import { SimpleQueryStatement } from "../../helper/Model";
import { debuglog, DebugLoggerFunction } from "node:util";
import { QueryCapture } from "tree-sitter";

let dlog: DebugLoggerFunction = debuglog("metric-gardener", (logger) => {
    dlog = logger;
});

export interface Accessor {
    name: string;
    namespaces: FullyQTN[];
    filePath: string;
    returnType: string;
}

export abstract class AbstractCollector {
    protected abstract getAccessorsQuery(): string;

    getAccessors(
        parseFile: ParseFile,
        namespacesOfFile: Map<string, FullyQTN>
    ): Map<string, Accessor[]> {
        const publicAccessors = new Map<string, Accessor[]>();

        if (this.getAccessorsQuery()) {
            const tree = TreeParser.getParseTree(parseFile);

            const queryBuilder = new QueryBuilder(parseFile.fileExtension, tree);
            queryBuilder.setStatements([new SimpleQueryStatement(this.getAccessorsQuery())]);

            const publicAccessorsQuery = queryBuilder.build();
            let publicAccessorsCaptures: QueryCapture[] = [];
            if (publicAccessorsQuery !== undefined) {
                publicAccessorsCaptures = publicAccessorsQuery.captures(tree.rootNode);
            }

            const accessorsTextCaptures: {
                name: string;
                text: string;
                usageType?: string;
                source?: string;
            }[] = formatCaptures(tree, publicAccessorsCaptures);

            dlog("public accessors captures", accessorsTextCaptures);

            // first index must be the return type
            // second index must be the accessor name
            for (let index = 0; index < accessorsTextCaptures.length; index += 1) {
                const publicAccessor: Accessor = {
                    name: "",
                    namespaces: Array.from(namespacesOfFile.values()),
                    filePath: parseFile.filePath,
                    returnType: accessorsTextCaptures[index].text,
                };

                index++;
                publicAccessor.name = accessorsTextCaptures[index].text;

                const accessorsByName = publicAccessors.get(publicAccessor.name);
                if (accessorsByName === undefined) {
                    publicAccessors.set(publicAccessor.name, []);
                }

                // TODO prevent duplicate adds by checking already existing combinations
                //  of name, returnType, and filepath
                publicAccessors.get(publicAccessor.name)?.push(publicAccessor);
            }
        }

        return publicAccessors;
    }
}
