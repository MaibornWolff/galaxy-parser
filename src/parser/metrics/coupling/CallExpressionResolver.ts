import { Relationship } from "../Metric";
import { Accessor } from "../../resolver/callExpressions/AbstractCollector";
import { UnresolvedCallExpression } from "../../resolver/typeUsages/AbstractCollector";
import { FullyQTN } from "../../resolver/fullyQualifiedTypeNames/AbstractCollector";

export function getAdditionalRelationships(
    tree: Map<string, Relationship[]>,
    unresolvedCallExpressions: Map<string, UnresolvedCallExpression[]>,
    publicAccessors: Map<string, Accessor[]>,
    alreadyAddedRelationships: Set<string>
) {
    let additionalRelationships: Relationship[] = [];

    for (const [filePath, callExpressions] of unresolvedCallExpressions) {
        const fileDependencies = [...(tree.get(filePath) ?? [])];
        if (!fileDependencies) {
            continue;
        }

        console.log("RESOLVING:", fileDependencies);

        const processedPublicAccessors = new Set<string>();

        const fileAdditionalRelationships: Relationship[] = [];

        for (const callExpression of callExpressions) {
            if (processedPublicAccessors.has(callExpression.name)) {
                continue;
            }
            processedPublicAccessors.add(callExpression.name);

            const qualifiedNameParts = callExpression.name.split(callExpression.namespaceDelimiter);
            if (callExpression.variableNameIncluded) {
                qualifiedNameParts.shift();
            }

            for (let namePart of qualifiedNameParts) {
                // remove save call character
                if (namePart.endsWith("?")) {
                    namePart = namePart.substring(0, namePart.length - 1);
                }

                const publicAccessorsOfPrefix = publicAccessors.get(namePart);
                if (publicAccessorsOfPrefix === undefined) {
                    continue;
                }

                const clonedAccessors = [...publicAccessorsOfPrefix];
                console.log(
                    "CLONED_ACCESSORS",
                    namePart,
                    filePath,
                    clonedAccessors.length,
                    fileDependencies
                );
                console.log(clonedAccessors);

                let added;

                for (const accessor of clonedAccessors) {
                    if (accessor.namespaces.length === 0) {
                        continue;
                    }

                    // Warning in case of Accessor Conflicts:
                    // In case that multiple accessors would fit the call expression,
                    // add both of them to not lose the correct dependency.
                    // this might lead to higher coupling values,
                    // but it is not that crucial as shown in the master thesis related to MetricGardener
                    for (const namespace of accessor.namespaces) {
                        const fullyQualifiedNameCandidate =
                            namespace.namespace +
                            namespace.namespaceDelimiter +
                            namespace.className;

                        console.log("\n\n", accessor, " -- ", fullyQualifiedNameCandidate);

                        // FirstAccessor is a property or method
                        // The type of myVariable must be an already added dependency of the current base type/class (filePath),
                        // so that subsequent method calls or attribute accesses can be resolved.
                        // Example:
                        // myVariable.FirstAccessor.SecondAccessor.ThirdAccessor
                        const baseDependency = fileDependencies.find((dependency) => {
                            return dependency.toNamespace === fullyQualifiedNameCandidate;
                        });

                        // In case of chained accesses, look in dependencies added for previous chain elements:
                        // e.g. look up already added dependency of return type of FirstAccessor to resolve SecondAccessor
                        // myVariable.FirstAccessor.SecondAccessor.ThirdAccessor
                        const callExpressionDependency = fileAdditionalRelationships.find(
                            (dependency) => {
                                return dependency.toNamespace === fullyQualifiedNameCandidate;
                            }
                        );

                        if (baseDependency !== undefined) {
                            added = resolveAccessorReturnType(
                                baseDependency,
                                accessor,
                                namespace,
                                tree,
                                fileAdditionalRelationships,
                                alreadyAddedRelationships
                            );
                        } else if (callExpressionDependency !== undefined) {
                            added = resolveAccessorReturnType(
                                callExpressionDependency,
                                accessor,
                                namespace,
                                tree,
                                fileAdditionalRelationships,
                                alreadyAddedRelationships
                            );
                        }

                        if (added) {
                            // Iterate only for the first namespace of an added accessor:
                            // One namespace of the accessor is fitting and
                            // the corresponding accessor was added as a dependency
                            // It seems not necessary to check the other namespaces of the accessor
                            break;
                        }
                    }
                }
            }
        }

        additionalRelationships = additionalRelationships.concat(fileAdditionalRelationships);
    }

    return additionalRelationships;
}

function resolveAccessorReturnType(
    matchingDependency: Relationship,
    accessor: Accessor,
    namespace: FullyQTN,
    tree: Map<string, Relationship[]>,
    additionalRelationships: Relationship[],
    alreadyAddedRelationships: Set<string>
) {
    // TODO resolve return type (generics, etc.)
    console.log(
        "Lookup Status: ",
        !!matchingDependency,
        " -> check return type add: ",
        accessor.returnType,
        matchingDependency,
        "\n\n"
    );

    const accessorFileDependencies = tree.get(namespace.source) ?? [];
    console.log("namespace.source", namespace.source);
    console.log("accessorFileDependencies", accessorFileDependencies);
    for (const accessorFileDependency of accessorFileDependencies) {
        // TODO Imagine that returnType is MyTypeNumberOne
        //  and toClassName MyType
        //  This would lead to a wrong dependency
        // Substring search using includes() is used because Return Types can look very differently:
        // Collection<MyTypeNumberOne>
        // Map<string, MyTypeNumberOne>
        // etc.
        if (accessor.returnType.includes(accessorFileDependency.toClassName)) {
            const uniqueId = accessorFileDependency.toNamespace + matchingDependency.fromNamespace;

            if (alreadyAddedRelationships.has(uniqueId)) {
                console.log(
                    "SKIP ADDING RETURN TYPE: ",
                    accessor.returnType,
                    "already added: ",
                    accessorFileDependency,
                    alreadyAddedRelationships.has(uniqueId),
                    "\n\n"
                );
                continue;
            }

            if (matchingDependency.fromNamespace !== accessorFileDependency.toNamespace) {
                alreadyAddedRelationships.add(uniqueId);

                const dependencyClone: Relationship = {
                    fromNamespace: matchingDependency.fromNamespace,
                    fromSource: matchingDependency.fromSource,
                    implementsCount: 0 /* TODO set right count or remove this property because it is used for testing purposes only */,
                    toClassName: accessorFileDependency.toClassName,
                    fromClassName: accessorFileDependency.toClassName,
                    toNamespace: accessorFileDependency.toNamespace,
                    toSource: accessorFileDependency.toSource,
                    usageType: "usage",
                };

                console.log("ADD RETURN TYPE: ", accessor.returnType, dependencyClone, "\n\n");
                return additionalRelationships.push(dependencyClone);
            }

            return 0;
        }
    }

    return 0;
}
