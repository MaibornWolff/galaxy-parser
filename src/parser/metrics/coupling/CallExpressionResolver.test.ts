import { describe, expect, it } from "vitest";
import { Relationship } from "../Metric";
import { getAdditionalRelationships } from "./CallExpressionResolver";
import { UnresolvedCallExpression } from "../../resolver/typeUsages/AbstractCollector";
import { Accessor } from "../../resolver/callExpressions/AbstractCollector";

describe("CallExpressionResolver", () => {
    describe("resolves call expressions and retrieves additional and transitive relationships", () => {
        it("when call expressions with save calls, public accessors and the right dependencies are given", () => {
            const firstItem: Relationship = {
                fromNamespace: "FirstItemNamespace.FirstItem",
                fromSource: "FirstItem",
                toSource: "SecondItem",
                toNamespace: "SecondItemNamespace.SecondItem",
                toClassName: "SecondItem",
                usageType: "usage",
            } as Relationship;

            const secondItem: Relationship = {
                fromNamespace: "SecondItemNamespace.SecondItem",
                fromSource: "SecondItem",
                toSource: "ThirdItem",
                toNamespace: "ThirdItemNamespace.ThirdItem",
                toClassName: "ThirdItem",
                usageType: "usage",
            } as Relationship;

            const thirdItem: Relationship = {
                fromNamespace: "ThirdItemNamespace.ThirdItem",
                fromSource: "ThirdItem",
                toSource: "FourthItem",
                toNamespace: "FourthItemNamespace.FourthItem",
                toClassName: "FourthItem",
                usageType: "usage",
            } as Relationship;

            const dependencyTree = new Map<string, Relationship[]>();
            dependencyTree.set(firstItem.fromSource, [firstItem]);
            dependencyTree.set(secondItem.fromSource, [secondItem]);
            dependencyTree.set(thirdItem.fromSource, [thirdItem]);
            dependencyTree.set(thirdItem.toSource, []);

            const callExpression1: UnresolvedCallExpression = {
                name: "myVariable?.UnknownAccessor?.AccessorInSecondItem?.AccessorInThirdItem?",
                namespaceDelimiter: ".",
                variableNameIncluded: true,
            };

            const callExpression2: UnresolvedCallExpression = {
                name: "myVariable.AccessorInSecondItem",
                namespaceDelimiter: ".",
                variableNameIncluded: true,
            };

            const unresolvedCallExpressions = new Map<string, UnresolvedCallExpression[]>();
            unresolvedCallExpressions.set(firstItem.fromSource, [callExpression1, callExpression2]);

            const accessor1: Accessor = {
                filePath: "SecondItem",
                name: "AccessorInSecondItem",
                namespaces: [
                    {
                        namespace: "SecondItemNamespace",
                        source: "SecondItem",
                        className: "SecondItem",
                        classType: "class",
                        namespaceDelimiter: ".",
                        implementedClasses: [],
                    },
                ],
                returnType: "ThirdItem",
            };

            const accessor2: Accessor = {
                filePath: "ThirdItem",
                name: "AccessorInThirdItem",
                namespaces: [
                    {
                        namespace: "ThirdItemNamespace",
                        source: "ThirdItem",
                        className: "ThirdItem",
                        classType: "class",
                        namespaceDelimiter: ".",
                        implementedClasses: [],
                    },
                ],
                returnType: "FourthItem",
            };

            const publicAccessors = new Map<string, Accessor[]>();
            publicAccessors.set(accessor1.name, [accessor1]);
            publicAccessors.set(accessor2.name, [accessor2]);

            const additionalRelationships = getAdditionalRelationships(
                dependencyTree,
                unresolvedCallExpressions,
                publicAccessors,
                new Set<string>(),
            );

            expect(additionalRelationships).toMatchSnapshot();
        });
    });
});
