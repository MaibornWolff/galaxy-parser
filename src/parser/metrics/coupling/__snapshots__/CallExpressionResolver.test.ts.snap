// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CallExpressionResolver > resolves call expressions and retrieves additional and transitive relationships > when call expressions with save calls, public accessors and the right dependencies are given 1`] = `
[
  {
    "fromClassName": "ThirdItem",
    "fromNamespace": "FirstItemNamespace.FirstItem",
    "fromSource": "FirstItem",
    "toClassName": "ThirdItem",
    "toNamespace": "ThirdItemNamespace.ThirdItem",
    "toSource": "ThirdItem",
    "usageType": "usage",
  },
  {
    "fromClassName": "FourthItem",
    "fromNamespace": "FirstItemNamespace.FirstItem",
    "fromSource": "FirstItem",
    "toClassName": "FourthItem",
    "toNamespace": "FourthItemNamespace.FourthItem",
    "toSource": "FourthItem",
    "usageType": "usage",
  },
]
`;
